Define the fallback case for service.name (#1269)
open-telemetry/opentelemetry-specification: specification/resource/semantic_conventions/README.md

Certain attribute groups in this document have a **Required** column. For these groups if any attribute from the particular group is present in the Resource then all attributes that are marked as Required MUST be also present in the Resource. However it is also valid if the entire attribute group is omitted (i.e. none of the attributes from the particular group are present even though some of them are marked as Required in this document).

## Attributes with Default Value

> The SDK MUST set a default value for the following attributes if they were not specified:

Reviewer: ```suggestion
The SDK MUST set a default value for the following attributes if they were not specified, even if the resource would otherwise not be present at all:
```

| `service.version` | string | The version string of the service API or implementation. | `2.0.0` | No |

**[1]:** MUST be the same for all instances of horizontally scaled services.

> Although this is a required attribute, if the value could not be specified SDKs

Reviewer 2: ```suggestion
Although this is a required attribute, if the value was not specified, SDKs
```

> MUST fallback to a language-specific name based on the running service.

Reviewer 2: Can we provide some guidance on how this could or should be done?

Author: This is actually language specific - but I added a mention of typically used components, such as process name, language or platform where the service is running.

> If this fallback fails, the name MUST default to `unknown_service`.

Reviewer 2: The part between those two semconv markers is generated, you'll have to add to the note in `semantic_conventions/resource/service.yaml` and then run the generator as described in `semantic_conventions/README.md`. The build is failing since the current markdown and the one that would be generated from the YAML don't match.

> If this fallback fails, the name MUST default to `unknown_service`.

Reviewer 2: Should we use a name that makes it more obvious this was not set by user code or an instrumentation but by the OpenTelemetry SDK itself?

Author: Open to suggestions :)

Reviewer 3: How about`$(telemetry.sdk.name)_service` or `openetemetry_service`?

Reviewer 3: After reading the other comments, I think that it would be best if we didn't offer a vague recommendation and then a fallback. A simpler rule will be: when the user does not set a required attribute, a determined value will be used (e.g., the two suggestions above). There are probably other semantic conventions for the attributes that might be used to infer "language-specific name based on the running service (typically inferred from process name, language, plaform, etc)."

Maybe just remove the "language-specific name" part and jump to a MUST and one of the suggestions above?

> | `service.instance.id` | string | The string ID of the service instance. [3] | `627cc493-f310-47de-96bd-71410b7dec09` | Yes |

Author: I'd suggest going ahead with this PR (as it has been standing for long time now) and follow up with #1034 right after (Priority 2) so we can either set it as well or remote it altogether.

> | `service.instance.id` | string | The string ID of the service instance. [3] | `627cc493-f310-47de-96bd-71410b7dec09` | Yes |

Reviewer: If we do that, we need to raise #1034 to P1 required-for-ga. If you think this is OK, I'm fine with merging as the PR is now. Note that doing it the other way round (resolving  #1034 first) would not result in the spec temporarily being in a contradictory state though.

> | `service.instance.id` | string | The string ID of the service instance. [3] | `627cc493-f310-47de-96bd-71410b7dec09` | Yes |

Reviewer: FYI @anuraaga

> | `service.instance.id` | string | The string ID of the service instance. [3] | `627cc493-f310-47de-96bd-71410b7dec09` | Yes |

Reviewer: A "quickfix" would be to make the instance-ID non-required.

> | `service.instance.id` | string | The string ID of the service instance. [3] | `627cc493-f310-47de-96bd-71410b7dec09` | Yes |

Reviewer 3: I am in favor of marking the instance-ID as non-required to let us merge this PR. Thanks for pointing out the inconsistency in #1034.

> | `service.instance.id` | string | The string ID of the service instance. [3] | `627cc493-f310-47de-96bd-71410b7dec09` | Yes |

Author: +1 for making instance-ID non-required (after all, there's the related issue). Will update, thanks!
